File: silencendo/src/ingestion/pipeline.ts
import * as fs from 'fs';
import * as path from 'path';
import * as url from 'url';
import pdfParse from 'pdf-parse';
import * as cheerio from 'cheerio';
import { KnowledgeSource } from '../sources/types';
import { Chunk, ChunkingOptions, IngestionResult } from './types';
import { randomUUID } from 'crypto';

export class IngestionPipeline {
  private chunkingOptions: ChunkingOptions;

  constructor(options?: Partial<ChunkingOptions>) {
    this.chunkingOptions = {
      chunkSize: options?.chunkSize || 1000,
      overlap: options?.overlap || 150,
    };
  }

  public async processSource(source: KnowledgeSource): Promise<IngestionResult> {
    try {
      let text: string;

      switch (source.type) {
        case 'file':
          text = await this.extractTextFromFile(source.path);
          break;
        case 'url':
          text = await this.extractTextFromUrl(source.path);
          break;
        default:
          throw new Error(`Unsupported source type: ${source.type}`);
      }

      const chunks = this.chunkText(text, source.id);
      return {
        chunks,
        sourceId: source.id,
        status: 'success',
      };
    } catch (error) {
      return {
        chunks: [],
        sourceId: source.id,
        status: 'error',
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  private async extractTextFromFile(filePath: string): Promise<string> {
    const resolvedPath = path.resolve(filePath);
    
    if (!fs.existsSync(resolvedPath)) {
      throw new Error(`File does not exist: ${resolvedPath}`);
    }

    const ext = path.extname(resolvedPath).toLowerCase();
    
    switch (ext) {
      case '.txt':
      case '.md':
        return fs.readFileSync(resolvedPath, 'utf-8');
      
      case '.pdf':
        const pdfData = await pdfParse(fs.readFileSync(resolvedPath));
        return pdfData.text;
      
      default:
        throw new Error(`Unsupported file type: ${ext}. Supported types: .txt, .md, .pdf`);
    }
  }

  private async extractTextFromUrl(url: string): Promise<string> {
    // For MVP, we'll simulate web scraping
    // In a real implementation, you would use a library like node-fetch or axios
    try {
      // Importing dynamically to avoid issues if not available
      const { default: fetch } = await import('node-fetch');
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      
      const html = await response.text();
      const $ = cheerio.load(html);
      
      // Remove script and style elements
      $('script, style, nav, footer, header').remove();
      
      // Extract text content
      return $('body').text().replace(/\s+/g, ' ').trim();
    } catch (error) {
      // If fetch is not available or fails, return a placeholder
      console.warn(`Could not fetch URL: ${url}. Using placeholder text.`);
      return `Content from ${url} (URL content could not be retrieved)`;
    }
  }

  private chunkText(text: string, sourceId: string): Chunk[] {
    const chunks: Chunk[] = [];
    const { chunkSize, overlap } = this.chunkingOptions;
    
    if (text.length <= chunkSize) {
      chunks.push({
        chunkId: randomUUID(),
        sourceId,
        text,
        metadata: {
          position: 0,
          length: text.length,
        },
      });
      return chunks;
    }

    // Split text into chunks with overlap
    let position = 0;
    while (position < text.length) {
      const chunkText = text.substring(position, position + chunkSize);
      chunks.push({
        chunkId: randomUUID(),
        sourceId,
        text: chunkText,
        metadata: {
          position,
          length: chunkText.length,
        },
      });

      // Move position by chunkSize minus overlap
      position += chunkSize - overlap;
      
      // If remaining text is less than or equal to overlap, we're done
      if (position >= text.length) {
        break;
      }
    }

    return chunks;
  }
}

File: silencendo/TMplatform/ingestion/pipeline.go
package ingestion

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"io/ioutil"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"silencendo/sources"
)

type IngestionPipeline struct {
	chunkingOptions ChunkingOptions
}

func NewIngestionPipeline(options *ChunkingOptions) *IngestionPipeline {
	opts := ChunkingOptions{
		ChunkSize: 1000,
		Overlap:   150,
	}

	if options != nil {
		if options.ChunkSize > 0 {
			opts.ChunkSize = options.ChunkSize
		}
		if options.Overlap >= 0 {
			opts.Overlap = options.Overlap
		}
	}

	return &IngestionPipeline{
		chunkingOptions: opts,
	}
}

func (ip *IngestionPipeline) ProcessSource(source sources.KnowledgeSource) (*IngestionResult, error) {
	var text string
	var err error

	switch source.Type {
	case "file":
		text, err = ip.extractTextFromFile(source.Path)
	case "url":
		text, err = ip.extractTextFromUrl(source.Path)
	default:
		err = errors.New("Unsupported source type: " + source.Type)
	}

	if err != nil {
		return &IngestionResult{
			Chunks:   []Chunk{},
			SourceID: source.ID,
			Status:   "error",
			Error:    err.Error(),
		}, nil
	}

	chunks := ip.chunkText(text, source.ID)
	return &IngestionResult{
		Chunks:   chunks,
		SourceID: source.ID,
		Status:   "success",
	}, nil
}

func (ip *IngestionPipeline) extractTextFromFile(filePath string) (string, error) {
	resolvedPath, err := filepath.Abs(filePath)
	if err != nil {
		return "", err
	}

	if _, err := os.Stat(resolvedPath); os.IsNotExist(err) {
		return "", errors.New("File does not exist: " + resolvedPath)
	}

	ext := strings.ToLower(filepath.Ext(resolvedPath))

	switch ext {
	case ".txt", ".md":
		content, err := ioutil.ReadFile(resolvedPath)
		if err != nil {
			return "", err
		}
		return string(content), nil

	case ".pdf":
		// Note: For a complete implementation, you would need a PDF parsing library
		// For now, returning a placeholder since we don't have a PDF parser in our imports
		return "PDF content would be extracted here", nil

	default:
		return "", errors.New("Unsupported file type: " + ext + ". Supported types: .txt, .md, .pdf")
	}
}

func (ip *IngestionPipeline) extractTextFromUrl(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "Content from " + url + " (URL content could not be retrieved)", nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "Content from " + url + " (URL content could not be retrieved)", nil
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return "Content from " + url + " (URL content could not be retrieved)", nil
	}

	// Remove script and style elements
	doc.Find("script, style, nav, footer, header").Each(func(i int, s *goquery.Selection) {
		s.Remove()
	})

	// Extract text content
	text := doc.Find("body").Text()
	return strings.Join(strings.Fields(text), " "), nil
}

func (ip *IngestionPipeline) chunkText(text string, sourceId string) []Chunk {
	var chunks []Chunk
	chunkSize := ip.chunkingOptions.ChunkSize
	overlap := ip.chunkingOptions.Overlap

	if len(text) <= chunkSize {
		chunk := Chunk{
			ChunkID:  generateUUID(),
			SourceID: sourceId,
			Text:     text,
		}
		chunk.Metadata.Position = 0
		chunk.Metadata.Length = len(text)
		chunks = append(chunks, chunk)
		return chunks
	}

	// Split text into chunks with overlap
	position := 0
	for position < len(text) {
		endPos := position + chunkSize
		if endPos > len(text) {
			endPos = len(text)
		}

		chunkText := text[position:endPos]
		chunk := Chunk{
			ChunkID:  generateUUID(),
			SourceID: sourceId,
			Text:     chunkText,
		}
		chunk.Metadata.Position = position
		chunk.Metadata.Length = len(chunkText)
		chunks = append(chunks, chunk)

		// Move position by chunkSize minus overlap
		position += chunkSize - overlap

		// If remaining text is less than or equal to overlap, we're done
		if position >= len(text) {
			break
		}
	}

	return chunks
}

// Helper function to generate a UUID
func generateUUID() string {
	bytes := make([]byte, 16)
	rand.Read(bytes)
	bytes[6] = (bytes[6] & 0x0f) | 0x40 // Version 4
	bytes[8] = (bytes[8] & 0x3f) | 0x80 // Variant 10
	return hex.EncodeToString(bytes)
}
